### PR TITLE
[HVG-588] Feature / proper sign status messages

### PR DESCRIPTION
**Pull request by styrelius**
## What?

Add better sign status messages rendered in `SignStatus` component, including a generic error message when something goes wrong when initiating the signing (i.e. you've already signed or something else is off):
1) New text key `CHECKOUT_SIGN_GENERIC_ERROR` when is `signUiState` is `'FAILED'`.
2) New `LinkToChat` component which renders a `TextButton` that opens Intercom chat when clicked. This is rendered with the generic error message.
3) A message when `signUiState` is `'STARTED_WITH_REDIRECT'` (which s true when you click sign in Norway and Denmark).
4) Some refactoring of `Checkout`, `Sign` and `Sign` components to make it clearer _to me_ what's going on there. 
**ADDITION:** New file `CheckoutSuccessRedirect.tsx` where I moved the Swedish success redirect and the `<TrackAction>` wrapping it.
5) New translations of some of the text keys used for Bank ID statuses. ~~A bit unclear to me though when some of these statuses are even returned/used.~~ 
**EDIT:** Also new names for the text keys so that it's clearer that they are only for _Swedish_  BankID, and  (after talking to backender Fredrik) I removed some of them that I deemed unnecessary.
6) Trigger the hiding and showing of the Intercom "launcher" (button) from `Checkout` component instead of from `OfferNew` component - this way the launcher is hidden even if you refresh the `/sign` page.
7) Clarify with the sign button text in the Swedish checkout that you have to sign with _mobile_  BankID.

_Please note_ that the possibility to cancel the signing session that is referenced in the ticked was too complicated to add at this point, since there is no support for it in the backend and thus we would have to make some fake cancelling or whatever.

## Why?

1) The main problem here has been that if the signing never starts due to some issues on the backend the user is not notified in any way - the sign button just shows a quick spinner and is then back to normal.
2) Another smaller issue is that if you are redirected to another url to sign (which you are in Norway and will be in Denmark) and this takes some time you won't get any information about what's going on.
3) Also some of the other status messages for the Swedish signing have implied that you could use BankID on your computer or on card to sign, which you really shouldn't do since you would not be able to log in to the Hedvig app without a mobile BankID on the device where you have the app installed. Overall we need to be clearer that this is _mobile_  BankID only for Sweden.

_Referenced ticket [HVG-588]_

## GIF's
**Swedish sign, successful after user cancelling sign in BankID app and then starting sign again:**
2020-12-18-SE-sign-cancelled-and-successful

**Swedish sign, failed:**
2020-12-14-SE-sign-failed

**Norwegian sign, successful:**
2020-12-14-NO-sign-successful

**Norwegian sign, failed:**
2020-12-14-NO-sign-failed

[HVG-588]: https://hedvig.atlassian.net/browse/HVG-588

### CODE REVIEW

**Review by gustaveen — COMMENTED** [review]
Good job Ellen, really nice with the chat link if the sign fails! 👌 Will test a bit locally to understand which errors are triggered

File: src/client/pages/OfferNew/Checkout/index.tsx [hunk 1]
```diff
@@ -225,11 +220,12 @@ export const Checkout: React.FC<Props> = ({
...
     }
   }, [member.data?.member, signStatus?.signState, variation])
+
   useScrollLock(visibilityState, scrollWrapper)
 
   const canInitiateSign = Boolean(
-    signUiState !== SignUiState.STARTED &&
-      signUiState !== SignUiState.STARTED_WITH_REDIRECT &&
+    signUiState !== 'STARTED' &&
+      signUiState !== 'STARTED_WITH_REDIRECT' &&
```
[the remark this answers is not in the document]
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/index.tsx:236]: I really disagree actually - I think it's so much easier to read with this kind of string union type. I changed it because I found it so hard to read the code with the enum we previously had. But if everyone else likes enums more I can of course change it.

File: src/translations/da_DK.json [hunk 2]
```diff
@@ -239,5 +239,12 @@
   "DETAILS_MODULE_TABLE_STUDENT_CELL_LABEL": "Student",
   "INVALID_SSN": "",
   "SSN_DOES_NOT_MATCH_BIRTH_DATE": "",
-  "INVALID_SSN_LENGTH": ""
+  "INVALID_SSN_LENGTH": "",
+  "CHECKOUT_SIGN_GENERIC_ERROR": "Noget gik galt. Prøv igen eller kontakt os.",
+  "CHECKOUT_SIGN_LINK_TO_OPEN_CHAT": "Chat med os",
```
[the remark this answers is not in the document]
Comment (styrelius) [reply src/translations/da_DK.json:244]: This is the Danish translation, the Norwegian looks exactly like your suggestion 🙃

[the remark this answers is not in the document]
Comment (gustaveen) [reply src/translations/da_DK.json:244]: Lol, I blame my bad morning brain 😂 Sorry for that! 🤦‍♂️

File: src/client/pages/OfferNew/Checkout/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (gustaveen) [reply src/client/pages/OfferNew/Checkout/index.tsx:236]: I like unions as well 👌, but I'm guessing there's a reason that we have enums in most cases like this. Since I'm no typescript wiz, what's you take @palmenhq @virrios

File: src/translations/da_DK.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (styrelius) [reply src/translations/da_DK.json:244]: Which I guess shows how skilled you are at Norwegian 🥇

[the remark this answers is not in the document]
Comment (gustaveen) [reply src/translations/da_DK.json:244]: Haha, or maybe not 😂

File: src/client/pages/OfferNew/Checkout/index.tsx [hunk 3]
```diff
@@ -201,6 +201,15 @@ export const Checkout: React.FC<Props> = ({
...
 
+  useEffect(() => {
+    if (visibilityState === VisibilityState.OPEN) {
+      Intercom('update', { hide_default_launcher: true })
+    }
+    if (visibilityState === VisibilityState.CLOSED) {
+      Intercom('update', { hide_default_launcher: false })
+    }
+  }, [visibilityState])
+
```
Comment (gustaveen) [inline src/client/pages/OfferNew/Checkout/index.tsx]: Awesome! 🙌

File: src/client/pages/OfferNew/Checkout/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/index.tsx:236]: I asked @palmenhq about that before, like more generally the difference between using enums and unions like these, and the way I understood it from his answers and from how enums works in these kinds of cases, they are completely interchangeable.

File: src/client/pages/OfferNew/Checkout/index.tsx (hunk 3 above)
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/index.tsx]: 😊 And this could easily be made into a custom hook if we want to hide the Intercom launcher somewhere else in the future.

File: src/client/pages/OfferNew/Checkout/index.tsx [hunk 4]
```diff
@@ -231,33 +231,34 @@ export const Checkout: React.FC<Props> = ({
...
 
-  const startSign = async () => {
+  const startSign = () => {
     if (!canInitiateSign) {
       return
     }
 
     const baseUrl = `${window.location.origin}/${locale}/new-member`
-    const result = await signQuotes({
+    signQuotes({
```
Comment (gustaveen) [inline src/client/pages/OfferNew/Checkout/index.tsx:248]: Just my opinion, I prefer `async/await` 🤓

Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/index.tsx:248]: Haha OK I totally respect that! 😄  The thing with using a `Promise` here was 1) that I could do a destructuring of `data` in a (to me) neat way and 2) that it enabled to easily add a `catch` without wrapping it all in a `try/catch` or having yet _another_ `if` to check if the request failed.

File: src/client/pages/OfferNew/Checkout/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (virrios) [reply src/client/pages/OfferNew/Checkout/index.tsx:236]: Fun tidbit since, we did this last week, (thought it was not necessary once we removed all danish type of Contracts and got removed)
Enums are iterable.  It was not at first possible for types but was later introduced, with this syntax
`
const contracts = ['travel', 'home', 'accident'] **as const;**
type IterableContractsType = typeof contracts[number]; // 'travel' | 'home' | 'accident'
`
I don't mind the enum, in terms of readability, but now its not really typesafe since its a numeric enum? And its actually passed as props in some components. 
I don't mind the readability of the union type either, but comparing when you are using it as to compare states I feel like its more natural to have it as a enum. 

I am not that strongly opinionated in this though 😄

File: src/client/pages/OfferNew/Checkout/index.tsx (hunk 4 above)
Comment (virrios) [reply src/client/pages/OfferNew/Checkout/index.tsx:248]: I have seen the usage of  the "then" to handle promises in our code but also await. I think we should be consistent
and the await syntax is more modern and a bit cleaner especially if have several steps.

File: src/client/pages/OfferNew/Checkout/SignStatus.tsx [hunk 5]
```diff
@@ -29,47 +49,58 @@ type Props = {
...
+        >
+          <StatusText>
+            {hasSigningError && (
+              <IconWrapper>
+                <WarningIcon color={colorsV3.gray700} />
+              </IconWrapper>
+            )}
+            {textKeys[signStatusTextKey]()}
+          </StatusText>
+          {hasSigningError && <LinkToChat />}
```
Comment (virrios) [inline src/client/pages/OfferNew/Checkout/SignStatus.tsx:101]: Awesome feature!fT! 🚀 🎆

File: src/client/pages/OfferNew/Checkout/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/index.tsx:236]: > I don't mind the enum, in terms of readability, but now its not really typesafe since its a numeric enum? And its actually passed as props in some components.

I'm not sure I understand fully @virrios - what is not really typesafe?

But yes, it's true that enums are iterable but in this case we're not using that (or any other of the special properties of the enum, as far as I can see) so here it's perfectly possible to use a union with strings. What I like is the readability when referencing those values, which in my view comes from the enum reference being more cluttered:
```javascript
  // less clutter :)
  signUiState !== 'STARTED'

  // more clutter
  signUiState !== SignUiState.STARTED
```
Especially confusing in these particular files where everything is called something with "sign" and "state" or "status" 😄

[the remark this answers is not in the document]
Comment (virrios) [reply src/client/pages/OfferNew/Checkout/index.tsx:236]: > I'm not sure I understand fully @virrios - what is not really typesafe?
> 
> But yes, it's true that enums are iterable but in this case we're not using that (or any other of the special properties of the enum, as far as I can see) so here it's perfectly possible to use a union with strings. What I like is the readability when referencing those values, which in my view comes from the enum reference being more cluttered:

Sry what I meant regarding that the enums are iterable was just to reference last week when we discussed the **strange** `as const` or `<const>['....']` syntax and I thought it was interesting and **fun** information **but not really applicable to what we are doing here 😄** 

**Regarding typesafeness** 
**This is numeric Enum because its auto incremented** 
`
enum Example {
a,
b
}
`
Its not type safe in the sense that If you pass it to a function, you can pass an arbitrary number as well. This type of enum is useful to use as a mask. However,  I saw that we use it as a property in another component and I think that you can pass just a number and it would not cause an error?. **This is my understanding** of numeric enums and  **correct me if I am wrong here!**.

So the union type would be more typesafe (or make the enum a string enum). 
So this was more to add to the general discussion.

** Regarding readability ***
I like enums and I think they are readable and not necessarily more cluttered.


Maybe we should decide when/when not to use enums/union? I like what you said in the meeting, so it looks like one person is writing the code.

[the remark this answers is not in the document]
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/index.tsx:229]: @virrios Ah, now I see! Interesting, didn't really know that about enums. But yes, let's have that discussion elsewhere then, and leave this with the changes I made since those are more type safe as Victor said?

File: src/client/pages/OfferNew/Checkout/index.tsx (hunk 4 above)
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/index.tsx:241]: I feel that both `Promises` and `async/await` they can both be useful for different cases and I'm not really sure that this is a case where you have to pick one? Although I generally love consistency 😀

File: src/client/pages/OfferNew/Checkout/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (palmenhq) [reply src/client/pages/OfferNew/Checkout/index.tsx:229]: Sorry i'm a bit late to the party here... Interesting point about the enums vs numbers, i had no idea but as it seems typescript is happy with the following code? Seems bad to me. I guess it's quite obvious that you shouldn't interchange enums with numbers but you really have a point here. For this case i think union types is good.

```ts
enum MyEnum { A, B }
const fn = (blah: MyEnum) => blah

fn(MyEnum.A) // OK
fn(0) // OK
fn(12) // OK??? like wat
```

File: src/client/pages/OfferNew/Checkout/index.tsx (hunk 4 above)
Comment (palmenhq) [reply src/client/pages/OfferNew/Checkout/index.tsx:241]: Just to chip in a third opinion i prefer `pipe` from rxjs ;)

Jokes aside i don't really have an opinion either. If we want to go towards a more functional style code base then maybe we should consider using `then` more (as generally it results in chaining functions rather than assigning tons of variables and hence state) but otherwise i agree `async`/`await` is a bit more readable. For this case imo it was nicer with `async`/`await` to be honest, but also we didn't actually catch if there was any errors. Adding a catch around this with nested ifs would decrease the prettiness quite a bit though, so i'd vote for keeping it with `then`s 

**tldr: keep it with `then` imo**

File: src/client/pages/OfferNew/Checkout/SignStatus.tsx [hunk 6]
```diff
@@ -1,75 +1,132 @@
...
+  signStatus: GraphQLSignStatus | null
+  signUiState: SignUiState
+  isLoading: boolean
+}
+
+enum StatusMessageTextKey {
+  REDIRECTING = 'CHECKOUT_SIGN_STARTED_WITH_REDIRECT',
+  GENERIC_ERR = 'CHECKOUT_SIGN_GENERIC_ERROR',
+  START_SIGNING = 'SE_BANKID_START_SIGN',
+  CERTIFICATE_ERR = 'SE_BANKID_CERTIFICATE_ERROR',
```
Comment (palmenhq) [inline src/client/pages/OfferNew/Checkout/SignStatus.tsx:21]: this is nit but i'm not sure i like shortening it to `err`. Err is i guess quite well accepted but i see no real reason here so we could just write it out imo

**Review by palmenhq — DISMISSED** [review]
THIS. IS. GREAT!! I had two more comments but you addressed them before i submitted my review haha. Very good job with this!

File: src/client/pages/OfferNew/Checkout/SignStatus.tsx (hunk 6 above)
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/SignStatus.tsx:21]: I agree, I was just annoyed when selecting all instances of `CERTIFICATE_ERROR` I also got that part of the `SE_BANKID_CERTIFICATE_ERROR` 😄 So I just wanted to differentiate them in some way, but I actually don't like `err` either. Can change it!

**Review by virrios — DISMISSED** [review]
Awesome job, and a super neat feature, for sure improve the user experience when signing!

**Comment by gustaveen** [comment]
Boom! Well done @styrelius 🍾 🎉

**Review by gustaveen — APPROVED** [review]
🚢
